### PR TITLE
🎉 New source: MongoDb ported to java

### LINKED ISSUE #3428: Port MongoDB Source to Java
## Tell us about the new integration you’d like to have
MongoDB is a critical source to support. Our current connector was contributed by a user. However, while the implementation is generally high quality, it is written in Ruby, and the Airbyte Core team's proficiencies are Java & Python. This means that we are much slower to implement features & bugfixes due to a lack of proficiency in Ruby. So we'd like to port the connector over to one of our core languages in order to offer better SLA & support.

## Describe the alternative you are considering or using
Continue to use current Ruby-based connector

## Implementation:
test container to use:
https://www.testcontainers.org/modules/databases/mongodb/

Todo:
1. Investigate possibility of using JDBC driver for mongo db (https://www.unityjdbc.com/mongojdbc/mongo_jdbc.php), but this seems to have the only paid jdbc driver which is not an option for us. Another option to check https://docs.mongodb.com/datalake/tutorial/jdbc-driver/ and  https://search.maven.org/search?q=a:mongodb-jdbc, but not sure how supportable it this driver for all DBs, this seems to be a kinda specific for Atlas Data Lake
In case of possible use of jdbc
2. Generate new connector and implement connections. 
3. Create unit test 
4. Integration test
5. Comprehensive:
- Investigate and prepare use cases:(https://docs.mongodb.com/manual/reference/bson-types/)
- Implement comprehensive test. 
6: Documentation, Prepare pull request, pass all checks. 
in case of need to implement with mongo native driver - :-(

use existing mongo source

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200367912513076/1200368131959422) by [Unito](https://www.unito.io)

Notes
It seems like the JDBC driver provided by unityjdbc is paid. So we have the same case here as it was for BigQuery. @DoNotPanicUA is currently working on db sources refactoring and implementation to make core better for such cases. So there is no value to start working on this ticket until the #4024 and #1876 are not completed. Then we would also need to support non-jdbc tests basics.
Aa this is non JDBC and even non SQL DB additional work in core part would be also required

### PR DESCRIPTION
**Pull request by irynakruk**
## What
Port MongoDB Source to Java

## How
Implemented with mongodb native java driver.
Added abstraction `AbstractDbSource` upon `AbstractRelationalDbSource`,  so now all non jdbc and non sql sources can depend on it.

## Recommended reading order
1. `MongoDbSource.java`
2. `MongoDatabase.java`
3. 

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector added to connector index like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
   
</p>
</details>

### CODE REVIEW

**Review by etsybaev — APPROVED** [review]
Please also make sure you have added all docs as per the checklist. 
 Connector's README.md
 docs/SUMMARY.md
 docs/integrations/<source or destination>/<name>.md including changelog. See changelog example
 docs/integrations/README.md
 airbyte-integrations/builds.md

File: airbyte-integrations/connectors/source-mongodb-new/CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# Changelog
+
+## 0.1.0
+Initial Release.
```
[the remark this answers is not in the document]
Comment (irynakruk) [reply airbyte-integrations/connectors/source-mongodb-new/CHANGELOG.md]: Thanks, added CHANGELOG section to README

File: airbyte-integrations/connectors/source-mongodb-new/README.md [hunk 2]
```diff
@@ -0,0 +1,4 @@
+# MongoDb Source
+
+## Documentation
+* [User Documentation](https://docs.airbyte.io/integrations/sources/mongodb-new)
```
[the remark this answers is not in the document]
Comment (irynakruk) [reply airbyte-integrations/connectors/source-mongodb-new/README.md]: Thanks, added

File: airbyte-db/lib/src/main/java/io/airbyte/db/SqlDatabase.java [hunk 3]
```diff
@@ -27,29 +27,10 @@
 import com.fasterxml.jackson.databind.JsonNode;
 import java.util.stream.Stream;
 
-public abstract class SqlDatabase implements AutoCloseable {
-
-  private JsonNode sourceConfig;
-  private JsonNode databaseConfig;
+public abstract class SqlDatabase extends NoSqlDatabase {
```
Comment (DoNotPanicUA) [inline airbyte-db/lib/src/main/java/io/airbyte/db/SqlDatabase.java:30]: Looks like we have a naming conflict here. `SqlDatabase` should not be an extension of `NoSqlDatabase`. 
As NoSqlDatabase doesn't contain any `no sql` logic, I propose to rename `NoSqlDatabase`.
For example, `Database`, `AbstractDatabase` or something like that.

File: airbyte-integrations/connectors/source-mongodb-new/build.gradle [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    mainClass = 'io.airbyte.integrations.source.mongodb.MongoDbSource'
+}
+
+dependencies {
+    implementation project(':airbyte-db:lib')
+    implementation project(':airbyte-integrations:bases:base-java')
+    implementation project(':airbyte-protocol:models')
+    implementation files(project(':airbyte-integrations:bases:base-java').airbyteDocker.outputs)
+    implementation project(':airbyte-integrations:connectors:source-relational-db')
+    implementation group: 'org.mongodb', name: 'mongo-java-driver', version: '3.12.8'
```
Comment (DoNotPanicUA) [inline airbyte-integrations/connectors/source-mongodb-new/build.gradle:17]: Do we really need the driver here? Is there something outside `airbyte-db` and `MongodbDatabase`?

**Review by DoNotPanicUA — APPROVED** [review]

File: airbyte-integrations/connectors/source-mongodb-new/build.gradle (hunk 4 above)
Comment (irynakruk) [reply airbyte-integrations/connectors/source-mongodb-new/build.gradle:17]: Yes, there is MongoCollection, BsonType and Document classes from mongo

File: airbyte-integrations/connectors/source-mongodb-new/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/MongoDbSourceAcceptanceTest.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+            .withCursorField(List.of("_id"))
+            .withStream(CatalogHelpers.createAirbyteStream(
+                    "test.acceptance_test",
+                    Field.of("_id", JsonSchemaPrimitive.STRING),
+                    Field.of("id", JsonSchemaPrimitive.STRING),
+                    Field.of("name", JsonSchemaPrimitive.STRING))
+                .withSupportedSyncModes(Lists.newArrayList(SyncMode.INCREMENTAL))
+                .withDefaultCursorField(List.of("_id")))));
+  }
+//
```
Comment (tuliren) [inline airbyte-integrations/connectors/source-mongodb-new/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/MongoDbSourceAcceptanceTest.java:123]: Nitpick.

```suggestion

```

**Review by tuliren — APPROVED** [review]
Looks great!

One minor issue about the name. `source-mongodb-new` seems informal. Can it be renamed to something like `source-mongodb-v2`?

@sherifnada, FYI.

File: airbyte-config/init/src/main/resources/seed/source_definitions.yaml [hunk 6]
```diff
@@ -458,3 +458,9 @@
   dockerRepository: airbyte/source-amazon-ads
   dockerImageTag: 0.1.0
   documentationUrl: https://docs.airbyte.io/integrations/sources/amazon-ads
+- sourceDefinitionId: b2e713cd-cc36-4c0a-b5bd-b47cb8a0561e
```
Comment (sherifnada) [inline airbyte-config/init/src/main/resources/seed/source_definitions.yaml:461]: lets remove the old mongo db implementation, having two implementations is confusing to the user

File: airbyte-integrations/builds.md [hunk 7]
```diff
@@ -49,6 +49,7 @@
 | Microsoft Teams   | [![source-microsoft-teams](https://img.shields.io/endpoint?url=https%3A%2F%2Fdnsgjos7lj2fu.cloudfront.net%2Ftests%2Fsummary%2Fsource-microsoft-teams%2Fbadge.json)](https://dnsgjos7lj2fu.cloudfront.net/tests/summary/source-microsoft-teams) |
 | Mixpanel   | [![source-mixpanel](https://img.shields.io/endpoint?url=https%3A%2F%2Fdnsgjos7lj2fu.cloudfront.net%2Ftests%2Fsummary%2Fsource-mixpanel%2Fbadge.json)](https://dnsgjos7lj2fu.cloudfront.net/tests/summary/source-mixpanel) |
 | Mongo DB   | [![source-mongodb](https://img.shields.io/endpoint?url=https%3A%2F%2Fdnsgjos7lj2fu.cloudfront.net%2Ftests%2Fsummary%2Fsource-mongodb%2Fbadge.json)](https://dnsgjos7lj2fu.cloudfront.net/tests/summary/source-mongodb) |
+| Mongo DB new  | [![source-mongodb-new](https://img.shields.io/endpoint?url=https%3A%2F%2Fdnsgjos7lj2fu.cloudfront.net%2Ftests%2Fsummary%2Fsource-mongodb-new%2Fbadge.json)](https://dnsgjos7lj2fu.cloudfront.net/tests/summary/source-mongodb-new) |
```
Comment (sherifnada) [inline airbyte-integrations/builds.md:52]: Externally, this should just be called `Mongo DB` and should replace the existing mongo implementation.

File: airbyte-integrations/connectors/source-mongodb-new/src/main/resources/spec.json [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+      },
+      "auth_source": {
+        "title": "Authentication source",
+        "type": "string",
+        "description": "Authentication source where user information is stored",
+        "default": "admin",
+        "examples": ["admin"],
+        "order": 5
+      },
+      "replica_set": {
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-mongodb-new/src/main/resources/spec.json:54]: can this connect to both a replica set _and_ a single node mongo db? 

Can you test it using the mongodb atlas credentials in lastpass?

File: docs/integrations/README.md [hunk 9]
```diff
@@ -65,6 +65,7 @@ Airbyte uses a grading system for connectors to help users understand what to ex
 |[Microsoft Teams](./sources/microsoft-teams.md)| Certified |
 |[Mixpanel](./sources/mixpanel.md)| Beta |
 |[Mongo DB](./sources/mongodb.md)| Alpha |
+|[Mongo DB new](./sources/mongodb-new.md)| Alpha |
```
Comment (sherifnada) [inline docs/integrations/README.md:68]: there should be only one mongo db, also it should be in beta

File: airbyte-config/init/src/main/resources/seed/source_definitions.yaml (hunk 6 above)
Comment (irynakruk) [reply airbyte-config/init/src/main/resources/seed/source_definitions.yaml:461]: I removed old implementation from source definitions.

File: docs/integrations/README.md (hunk 9 above)
Comment (irynakruk) [reply docs/integrations/README.md]: Done

File: airbyte-integrations/builds.md (hunk 7 above)
Comment (irynakruk) [reply airbyte-integrations/builds.md]: Removed old `| Mongo DB   |` and renamed a new one to `| Mongo DB   |` accordingly

File: airbyte-integrations/connectors/source-mongodb-new/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/MongoDbSourceAcceptanceTest.java (hunk 5 above)
Comment (irynakruk) [reply airbyte-integrations/connectors/source-mongodb-new/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/MongoDbSourceAcceptanceTest.java]: Thanks, removed

File: airbyte-integrations/connectors/source-mongodb-new/src/main/resources/spec.json (hunk 8 above)
Comment (irynakruk) [reply airbyte-integrations/connectors/source-mongodb-new/src/main/resources/spec.json]: Sure, working on it

Comment (irynakruk) [reply airbyte-integrations/connectors/source-mongodb-new/src/main/resources/spec.json]: @sherifnada I've updated spec.json and connection param to connect to Mongodb Atlas as it uses `mongodb+srv` protocol.
So now it is possible to connect to Standalone MongoDb instance, Replica Set and MongoDb Atlas Cluster as well.
I've tested it using the mongodb atlas credentials, works good.

**Review by sherifnada — APPROVED** [review]
nice!

**Pull request merged**